### PR TITLE
Auto uppercase middle name

### LINKED ISSUE #42: auto_uppercase for 'sólar í dag í reykjavík"
Tokenizing the phrase 'hver er hæð sólar í dag í reykjavík' with `auto_uppercase=True` results in 'Sólar Í Dag Í Reykjavík' being interpreted as a single name token (`TOK.PERSON`).

image
image

This also happens when 'í reykjavík' is omitted (resulting in 'Sólar Í Dag').
('hæð' probably shouldn't be capitalized either in this case).

### PR DESCRIPTION
**Pull request by sultur**
Changes to middle name abbreviation functionality when auto_uppercasing.

Interprets words in `MIDDLE_NAME_ABBREV` as middle name abbreviations (adding them to the name) and words in `NOT_NAME_ABBREVS` as abbreviations **only if they are followed by a period (".")**.

Fixes #42 (at least partially, 'sólar' and 'dag' are still capitalized along with "hæð', but 'sólar í dag í reykjavík' doesn't get interpreted as a single name).

### CODE REVIEW

File: src/reynir/bintokenizer.py [hunk 1]
```diff
@@ -412,6 +415,7 @@ def all_except_suffix(s: str) -> str:
     ("in", "a", "an", "for", "and", "the", "for", "on", "of")
 )
 
+SENTENCE_END_START_TOKENS: FrozenSet[int] = frozenset((TOK.S_END, TOK.S_BEGIN))
```
Comment (vthorsteinsson) [inline src/reynir/bintokenizer.py:418]: Nitpick: A better name would be SENTENCE_DELIMITER_TOKENS

File: src/reynir/bintokenizer.py [hunk 2]
```diff
@@ -1047,7 +1060,11 @@ def parse_phrases_2(
         at_sentence_start = False
 
         while True:
-            next_token = next(token_stream)
+            next_token = (
+                following_tokens.popleft()
+                if len(following_tokens)
```
Comment (vthorsteinsson) [inline src/reynir/bintokenizer.py:1065]: Interesting solution! One question: is `len(following_tokens)` needed or can you do just `if following_tokens`?

File: src/reynir/bintokenizer.py [hunk 3]
```diff
@@ -1204,7 +1221,7 @@ def unknown_surname(tok: Tok) -> bool:
                 """ Check for unknown (non-Icelandic) surnames """
                 # Accept (most) upper case words as a surnames
                 if auto_uppercase:
-                    if tok.txt in UPPER_CASE_ABBREVS:
+                    if tok.txt in MIDDLE_NAME_ABBREVS:
```
Comment (vthorsteinsson) [inline src/reynir/bintokenizer.py:1224]: Might be good to update the comment below to accurately describe what's happening, because it's starting to get really complicated. In other words, let's have mercy on future programmers who will be trying to understand this code ;-)

File: src/reynir/bintokenizer.py [hunk 4]
```diff
@@ -1228,24 +1245,27 @@ def given_names_or_middle_abbrev(
...
-                wrd = tok.txt
-                if len(wrd) > 2 or wrd[0].islower():
-                    if wrd not in FOREIGN_MIDDLE_NAME_SET:
-                        # Accept "Thomas de Broglie", "Ruud van Nistelrooy"
-                        return None
-                # One or two letters, capitalized: accept as middle name abbrev,
-                # all genders and cases possible
+                if auto_uppercase and wrd in MIDDLE_NAME_ABBREVS:
+                    wrd = wrd.capitalize()
+                if len(wrd) > 3 or (
```
Comment (vthorsteinsson) [inline src/reynir/bintokenizer.py:1262]: Why the change to len > 3 from len > 2? A comment here would be good.

File: src/reynir/bintokenizer.py [hunk 5]
```diff
@@ -1287,6 +1307,37 @@ def compatible(pn: PersonNameTuple, npn: PersonNameTuple) -> bool:
...
+                    # Deal with lowercase middle name abbreviations when auto uppercasing
+                    if auto_uppercase and next_token.txt in MIDDLE_NAME_ABBREVS:
+                        next_next_token = next(token_stream)
+                        following_tokens.append(next_next_token)
+
+                        # Check if next token is a period
+                        if next_next_token.punctuation == ".":
+                            # Join to middle name abbreviation
+                            next_token = next_token.concatenate(next_next_token)
+                            following_tokens.popleft()
```
Comment (vthorsteinsson) [inline src/reynir/bintokenizer.py:1320]: Are you sure that you want to pop from the left side, not from the right side (where you previously appended next_next_token)?

File: src/reynir/bintokenizer.py [hunk 6]
```diff
@@ -1287,6 +1307,37 @@ def compatible(pn: PersonNameTuple, npn: PersonNameTuple) -> bool:
...
+
+                            try:
+                                # Remove sentence end/start tokens
+                                while (
+                                    next_next_token.punctuation == "."
+                                    or next_next_token.kind in SENTENCE_END_START_TOKENS
+                                ):
+                                    next_next_token = next(token_stream)
```
Comment (vthorsteinsson) [inline src/reynir/bintokenizer.py:1328]: Shouldn't you be appending to following_tokens here, in this inner while loop?

File: test/test_parse.py [hunk 7]
```diff
@@ -1789,6 +1789,45 @@ def test_names(r):
     assert s.tree is not None
     assert s.tree.persons == []
     assert s.tree.nouns == ["sýning", "bíó", "þriðjudagskvöld"]
+    s = r.parse_single("Ruud van Nistelrooy og Thomas de Broglie komu í heimsókn.")
+    assert "Thomas de Broglie" in s.tree.persons
+    # "Ruud van Nistelrooy" gets interpreted as an entity
+    # assert "Ruud van Nistelrooy" in s.tree.persons
```
Comment (vthorsteinsson) [inline test/test_parse.py:1795]: We could still check for 'Ruud van Nistelrooy' as a whole (entity) token in the test, I recommend doing that.

File: test/test_reynir.py [hunk 8]
```diff
@@ -317,13 +317,134 @@ def test_auto_uppercase():
...
     assert detokenize(s.tokens) == "hver er Eliza Reid"
     assert "Eliza Reid" in s.tree.persons
 
+    s = g.parse_single("hver er hæð jóns")
+    assert detokenize(s.tokens) == "hver er hæð Jóns"
+    assert "Jón" in s.tree.persons
+
+    s = g.parse_single("hver er hæð sólar í dag í reykjavík")
+    assert "Í" not in detokenize(s.tokens)
+    assert "Sólar Í Dag Í Reykjavík" not in s.tree.persons
```
Comment (vthorsteinsson) [inline test/test_reynir.py:339]: Could add further tests for handling of "sólar" and "dag", I know some of them might not be producing 100% correct results but at least having them here and eventually commented out would be good as a reminder.

File: src/reynir/bintokenizer.py (hunk 5 above)
Comment (sultur) [reply src/reynir/bintokenizer.py:1320]: Yes, I am really just using the deque as a queue ADT (as Python doesn't have a pure queue implementation in any library). next_next_token should always be at the front of the queue (or the left side of the deque) so popleft is intentional.

File: src/reynir/bintokenizer.py (hunk 6 above)
Comment (sultur) [reply src/reynir/bintokenizer.py:1328]: The reason I don't append to following_tokens is to remove the sentence delimiter tokens (`S_BEGIN`,`S_END`), which get added in the `annotate` phase before `parse_phrases_2` gets invoked.

I want to interpret e.g. the _k_ and _._ in _þorlákur k. sæmundarson_ as a single token without the `S_BEGIN` and `S_END` tokens. Maybe we want to keep the S_END token though if there are no tokens after the name?

File: src/reynir/bintokenizer.py (hunk 1 above)
Comment (sultur) [reply src/reynir/bintokenizer.py]: Fixed!

File: src/reynir/bintokenizer.py (hunk 2 above)
Comment (sultur) [reply src/reynir/bintokenizer.py]: Good point! len() isn't needed.

File: src/reynir/bintokenizer.py (hunk 3 above)
Comment (sultur) [reply src/reynir/bintokenizer.py]: I realized this part of the function wasn't needed so I simplified it. Now it only deals with foreign middle names (unknown words) so it's a bit more straight-forward.

File: src/reynir/bintokenizer.py (hunk 4 above)
Comment (sultur) [reply src/reynir/bintokenizer.py:1262]: Added commentary. The length was changed due to the following period being included in wrd if the abbreviation has a period. E.g. _th._ or _kr._ should be accepted as middle name abbreviations, but also shorter abbrevs such as _th_, _h._ or _b_

File: test/test_parse.py (hunk 7 above)
Comment (sultur) [reply test/test_parse.py]: Added entity check!

File: src/reynir/bintokenizer.py (hunk 6 above)
Comment (vthorsteinsson) [reply src/reynir/bintokenizer.py:1328]: We want to keep S_END, yes, it may be relied upon in subsequent passes.

File: src/reynir/bintokenizer.py [hunk 9]
```diff
@@ -1259,13 +1256,16 @@ def given_names_or_middle_abbrev(
                     return None
                 if auto_uppercase and wrd in MIDDLE_NAME_ABBREVS:
                     wrd = wrd.capitalize()
+                # If wrd is longer than middle name abbrevs (possibly with following period)
+                # such as "th.", "kr" or "f."
```
Comment (vthorsteinsson) [inline src/reynir/bintokenizer.py:1260]: Still not convinced that this hould be len(wrd) > 3 and not len(wrd) > 2

File: src/reynir/bintokenizer.py (hunk 5 above)
Comment (sultur) [reply src/reynir/bintokenizer.py:1320]: On second thought, it is probably safer to use pop rather than popleft, good point!

File: src/reynir/bintokenizer.py (hunk 9 above)
Comment (sultur) [reply src/reynir/bintokenizer.py]: Yeah I found a few edge cases which confused the tokenizer due to that. I changed it to `len(wrd.rstrip(".")) > 2` and now everything seems to work!

File: src/reynir/bintokenizer.py (hunk 6 above)
Comment (sultur) [reply src/reynir/bintokenizer.py]: Fixed!
